File: epi_judge_java/epi/PhoneNumberMnemonic.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.EpiTestComparator;
import epi.test_framework.GenericTest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PhoneNumberMnemonic {
    private static final String[] MAPPING = {"0", "1", "ABC", "DEF", "GHI",
            "JKL", "MNO", "PQRS", "TUV", "WXYZ"};

    @EpiTest(testDataFile = "phone_number_mnemonic.tsv")
    public static List<String> phoneMnemonic(String phoneNumber) {
        List<String> result = new ArrayList<>();

        recur(phoneNumber, 0, new StringBuilder(), result);

        return result;
    }

    private static void recur(String phoneNumber, int index, StringBuilder sb, List<String> result) {
        if (sb.length() == phoneNumber.length()) {
            result.add(sb.toString());
        } else {
            int mappingIndex = phoneNumber.charAt(index) - '0';
            for (char c : MAPPING[mappingIndex].toCharArray()) {
                sb.append(c);
                recur(phoneNumber, index + 1, sb, result);
                sb.setLength(sb.length() - 1);
            }
        }
    }

    @EpiTestComparator
    public static boolean comp(List<String> expected, List<String> result) {
        if (result == null) {
            return false;
        }
        Collections.sort(expected);
        Collections.sort(result);
        return expected.equals(result);
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "PhoneNumberMnemonic.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/phone_number_mnemonic.cc
#include <string>
#include <vector>

#include "test_framework/generic_test.h"

using std::string;
using std::vector;
using std::array;

const array<string, 10> kMapping = {
        {"0", "1", "ABC", "DEF", "GHI", "JKL", "MNO", "PQRS", "TUV", "WXYZ"}};

vector<string> result;

void recur(const string& phone_number, int index, string& curr);

vector<string> PhoneMnemonic(const string& phone_number) {
    result.clear();

    string curr;
    recur(phone_number, 0, curr);

    return result;
}

void recur(const string& phone_number, int index, string& curr) {
    if (curr.size() == phone_number.size()) {
        result.emplace_back(curr);
        return;
    }

    for (auto symbol : kMapping[phone_number[index] - '0']) {
        curr += symbol;
        recur(phone_number, index + 1, curr);

        curr.pop_back();
    }
}

int main(int argc, char* argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"phone_number"};
    return GenericTestMain(args, "phone_number_mnemonic.cc",
                           "phone_number_mnemonic.tsv", &PhoneMnemonic,
                           UnorderedComparator{}, param_names);
}
